File: src/lib/resume-render.ts
import type { Resume } from "./types";

export interface RenderOptions {
  /** Overrides the resume summary with a role-specific one. */
  summary?: string;
  /** Overrides bullets per experience id, letting the tailor drop or reorder them. */
  bulletsByExperienceId?: Record<string, string[]>;
  /** Skills to surface first within each group. */
  prioritySkills?: string[];
  targetTitle?: string;
}

function formatDateRange(start: string, end: string): string {
  const label = (value: string) => {
    if (!value) return "";
    if (/^\d{4}-\d{2}$/.test(value)) {
      const [year, month] = value.split("-");
      const monthName = new Date(Number(year), Number(month) - 1, 1).toLocaleString("en-US", {
        month: "short",
      });
      return `${monthName} ${year}`;
    }
    return value;
  };
  const from = label(start);
  const to = label(end);
  if (from && to) return `${from} – ${to}`;
  return from || to;
}

function orderSkills(items: string[], priority: string[]): string[] {
  if (priority.length === 0) return items;
  const normalized = new Set(priority.map((item) => item.toLowerCase()));
  const promoted = items.filter((item) => normalized.has(item.toLowerCase()));
  const rest = items.filter((item) => !normalized.has(item.toLowerCase()));
  return [...promoted, ...rest];
}

export function renderResumeMarkdown(resume: Resume, options: RenderOptions = {}): string {
  const { basics } = resume;
  const lines: string[] = [];

  lines.push(`# ${basics.name}`);
  const title = options.targetTitle || basics.title;
  if (title) lines.push(`**${title}**`);

  const contact = [basics.location, basics.email, basics.phone].filter(Boolean).join(" · ");
  const links = basics.links
    .filter((link) => link.url)
    .map((link) => `[${link.label}](${link.url})`)
    .join(" · ");
  if (contact) lines.push("", contact);
  if (links) lines.push(links);

  const summary = options.summary || basics.summary;
  if (summary) lines.push("", "## Summary", "", summary);

  if (resume.skillGroups.length > 0) {
    lines.push("", "## Skills", "");
    for (const group of resume.skillGroups) {
      const items = orderSkills(group.items, options.prioritySkills ?? []);
      if (items.length > 0) lines.push(`**${group.label}:** ${items.join(", ")}`, "");
    }
  }

  if (resume.experience.length > 0) {
    lines.push("", "## Experience", "");
    for (const role of resume.experience) {
      lines.push(`### ${role.role} — ${role.company}`);
      const meta = [role.location, formatDateRange(role.start, role.end)].filter(Boolean).join(" · ");
      if (meta) lines.push(`*${meta}*`);
      lines.push("");

      const bullets = options.bulletsByExperienceId?.[role.id] ?? role.bullets;
      for (const bullet of bullets) lines.push(`- ${bullet}`);
      if (role.stack.length > 0) {
        lines.push("", `*Stack: ${orderSkills(role.stack, options.prioritySkills ?? []).join(", ")}*`);
      }
      lines.push("");
    }
  }

  if (resume.projects.length > 0) {
    lines.push("## Projects", "");
    for (const project of resume.projects) {
      const heading = project.url ? `[${project.name}](${project.url})` : project.name;
      lines.push(`### ${heading}`);
      lines.push("", project.description);
      if (project.stack.length > 0) lines.push("", `*Stack: ${project.stack.join(", ")}*`);
      lines.push("");
    }
  }

  if (resume.education.length > 0) {
    lines.push("## Education", "");
    for (const entry of resume.education) {
      const range = formatDateRange(entry.start, entry.end);
      lines.push(`**${entry.degree}** — ${entry.school}${range ? ` (${range})` : ""}`);
      if (entry.detail) lines.push("", entry.detail);
      lines.push("");
    }
  }

  if (resume.certifications.length > 0) {
    lines.push("## Certifications", "");
    for (const certification of resume.certifications) lines.push(`- ${certification}`);
    lines.push("");
  }

  return lines.join("\n").replace(/\n{3,}/g, "\n\n").trim();
}

const ESCAPES: Record<string, string> = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': "&quot;",
};

export function escapeHtml(value: string): string {
  return value.replace(/[&<>"]/g, (char) => ESCAPES[char]);
}

File: app/resume_render.py
"""Turn a Resume into markdown. Tailoring never invents rows."""

from __future__ import annotations

from app.models import Resume


def format_date_range(start: str, end: str) -> str:
    def label(value: str) -> str:
        if not value:
            return ""
        if len(value) == 7 and value[4] == "-":
            year, month = value.split("-")
            months = "Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec".split()
            try:
                return f"{months[int(month) - 1]} {year}"
            except (ValueError, IndexError):
                return value
        return value

    from_ = label(start)
    to = label(end)
    if from_ and to:
        return f"{from_} – {to}"
    return from_ or to


def order_skills(items: list[str], priority: list[str]) -> list[str]:
    if not priority:
        return items
    wanted = {item.lower() for item in priority}
    promoted = [item for item in items if item.lower() in wanted]
    rest = [item for item in items if item.lower() not in wanted]
    return promoted + rest


def render_resume_markdown(
    resume: Resume,
    *,
    summary: str | None = None,
    bullets_by_experience_id: dict[str, list[str]] | None = None,
    priority_skills: list[str] | None = None,
    target_title: str | None = None,
) -> str:
    basics = resume.basics
    lines: list[str] = [f"# {basics.name}"]
    title = target_title or basics.title
    if title:
        lines.append(f"**{title}**")

    contact = " · ".join(part for part in (basics.location, basics.email, basics.phone) if part)
    links = " · ".join(
        f"[{link.get('label', 'link')}]({link['url']})" for link in basics.links if link.get("url")
    )
    if contact:
        lines.extend(["", contact])
    if links:
        lines.append(links)

    summary_text = summary or basics.summary
    if summary_text:
        lines.extend(["", "## Summary", "", summary_text])

    if resume.skill_groups:
        lines.extend(["", "## Skills", ""])
        for group in resume.skill_groups:
            items = order_skills(group.items, priority_skills or [])
            if items:
                lines.extend([f"**{group.label}:** {', '.join(items)}", ""])

    if resume.experience:
        lines.extend(["", "## Experience", ""])
        for role in resume.experience:
            lines.append(f"### {role.role} — {role.company}")
            meta = " · ".join(part for part in (role.location, format_date_range(role.start, role.end)) if part)
            if meta:
                lines.append(f"*{meta}*")
            lines.append("")
            bullets = (bullets_by_experience_id or {}).get(role.id, role.bullets)
            for bullet in bullets:
                lines.append(f"- {bullet}")
            if role.stack:
                lines.extend(["", f"*Stack: {', '.join(order_skills(role.stack, priority_skills or []))}*"])
            lines.append("")

    if resume.projects:
        lines.extend(["## Projects", ""])
        for project in resume.projects:
            heading = f"[{project.name}]({project.url})" if project.url else project.name
            lines.extend([f"### {heading}", "", project.description])
            if project.stack:
                lines.extend(["", f"*Stack: {', '.join(project.stack)}*"])
            lines.append("")

    if resume.education:
        lines.extend(["## Education", ""])
        for entry in resume.education:
            span = format_date_range(entry.start, entry.end)
            suffix = f" ({span})" if span else ""
            lines.append(f"**{entry.degree}** — {entry.school}{suffix}")
            if entry.detail:
                lines.extend(["", entry.detail])
            lines.append("")

    if resume.certifications:
        lines.extend(["## Certifications", ""])
        for cert in resume.certifications:
            lines.append(f"- {cert}")
        lines.append("")

    text = "\n".join(lines)
    while "\n\n\n" in text:
        text = text.replace("\n\n\n", "\n\n")
    return text.strip()


def escape_html(value: str) -> str:
    return (
        value.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
    )
